plugin/loadbalance/README.md

```diff
@@ -19,7 +19,40 @@ implementations (like glibc) are particular about that.
 loadbalance [POLICY]
 ~~~
 
-* **POLICY** is how to balance. The default, and only option, is "round_robin".
+* **POLICY** is how to balance. Available options are "round_robin" and "weighted_round_robin". The default is "round_robin".
+
+The "round_robin" strategy randomizes the order of  A, AAAA, and MX records applying a uniform probability distribution.
+
+The "weighted_round_robin" policy assigns weight values to IPs to control the relative likelihood of particular IPs to be returned as top
+entry (A or AAAA record) in the answer. See [Wikipedia](https://en.wikipedia.org/wiki/Weighted_round_robin) about a generic
+description of weighted round robin load balancing strategy.
+
+Additional option required by "weighted_round_robin" policy:
+
+~~~
+loadbalance weighted_round_robin WEIGHTFILE
+~~~
+
+* **WEIGHTFILE** the weight file containing the weight values assigned to IPs. If the path is relative, the path from the **root** plugin will be prepended to it.
+
+The weight file is parsed line-by-line. If the first character in the line is '#' then the line is ignored (i.e. comment line). Otherwise, if the line contains a single word then it is a domain name. If the line contains two words then the first one is an IP and the second one is a weight value for that IP. The IPs are addresses for the domain name which is defined above them in the file. Multiple domain names can be specified in the same weight file. A weight value must be in the range [1,255].
+
+If the server receives a query for which no domain name is specified in the corresponding weight file then the answer is returned unmodified. If the domain name is specified in the weight file but the answer does not include the expected top IP (or there isn't any IP/weight pair specified for the particular domain name) then the answer is returned unmodified.
```

Comment:
Not really. The weighted-round-robin policy does not shuffle all the elements of the result lists (as opposed to the default round-robin). It only tries to move a particular IP to the first position in the list.  So, whenever it gets a new result list it does the following two steps: 

1.  it computes the "next expected IP" from the weighted IPs list 
2.  it checks if that particular IP is in the result list. If so, then it moves that IP to the first (top) position in the list.

The rest of the result list is not modified (i.e. not shuffled). The weights only determine how likely an IP will be returned as the first item in the result list.

Reply:
Thanks.  IMO the wording in the readme doesn't clearly capture that.

The behavior seems counterintuitive - that results will be sometimes weighted and sometimes not for the same query name and answer set.

It would seem more intuitive to only return the answer unaltered if the answer contains a record that is not defined in the weight file.  Otherwise, the answer records all have defined weights, and one can be selected to be the top/first record in the response.

An alternate approach would be to only return the answer unaltered if the answer does not contain any records defined in the weight file.  Otherwise, only records with defined weights are eligible to be selected for the top/first record in the response.  Additionally, when returning the answer "unaltered", it would probably be better to randomly select a top IP, as if all the records have equal weight (which in this case they do in a sense).

Yet another approach would be to define in the weight file _default_ weights to be used for IPs that do not have a weight defined.

Reply:
It makes sense, thanks for the suggestions! I changed the implementation to assign weight for every address record in the result list. If an address does not have a weight defined in the weight file then the default weight value "1" is assumed for it.  The answer is returned unaltered only if there isn't any address record.

The default weight value (1) is hardcoded currently as I'm not sure if it is really necessary to make that configurable (via the weight file).